### PR TITLE
Ci fix after moving to cpp17

### PR DESCRIPTION
**Pull request by AlekMosingiewicz**
Issue this pull request references: #454 
Attempt to update CI configuration for the build to pass with new C++ standard.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -5,10 +5,13 @@ addons:
   apt:
     sources: 
       - ubuntu-toolchain-r-test
+      - llvm-toolchain-trusty-5.0
+      - sourceline: 'deb http://apt.llvm.org/trusty/ llvm-toolchain-trusty-5.0 main'
+        key_url: 'https://apt.llvm.org/llvm-snapshot.gpg.key'
```
Comment (RobLoach) [inline .travis.yml]: As an alternative to the llvm snapshots, I've used the following before....

```
addons:
  apt:
    sources:
    - sourceline: 'ppa:ubuntu-toolchain-r/test'
    packages:
    - cmake
    - time
    - g++-7
    - gcc-7
    - binutils
    - libc6
    - libc6-amd64
    - libc6-dev
    - libc6-dbg
```

**Comment by lefticus** [comment]
@AlekMosingiewicz thank you for the hard work with this. Looks like I need to spend a few minutes with the fuzzy tests and windows builds.

I'm not sure what to say about the MacOS builds - apple makes it very hard to know what compiler they are actually using since they make up their own version numbers that don't match up with LLVM's

File: .travis.yml (hunk 1 above)
Comment (AlekMosingiewicz) [reply .travis.yml]: Travis pipeline for OSX uses XCode, which as far as I know has its own version of Clang, so I don't think this will help much.

**Comment by AlekMosingiewicz** [comment]
I'd turn off the OSX pipeline until Travis version of XCode is up to date.

**Comment by lefticus** [comment]
On another project I've been trying out other CI systems (circle) and will look soon at cirrus-ci. Travis is staying woefully out of date and making all of this harder than it needs to be. Maybe it makes sense to move primarily over to one of the others?

**Comment by AlekMosingiewicz** [comment]
I haven't had much time to work on the CI lately but I believe I had fixed the Travis pipeline by updating the GCC - the latest fail was some internal Travis error, this build has passed: 
https://travis-ci.org/ChaiScript/ChaiScript/builds/425304574
The latest failed build didn't include any changes in Travis configuration, it was an attempt to actually tidy AppVeyor.
AppVeyor itself needs a look at why the build is failing on Visual Studio in the first place.

**Comment by lefticus** [comment]
I tried and was unable to get this branch of Chai compiling with the latest Visual Studio yet.

**Comment by AlekMosingiewicz** [comment]
I can't get master branch to compile on my Visual Studio either. The only change to code in this branch is the preprocessor directive to fetch "variant" header file from the "experimental" folder ( I've read that this is where it's stored for some of the later versions of XCode, I might scrap this change ). Other than that I'm enforcing the C++17 standards in Cmake, this is how I eliminated most of the compile errors for XCode Clang ( the only one left is about missing "variant" header ).

**Comment by RobLoach** [comment]
I had to revert to https://github.com/ChaiScript/ChaiScript/tree/ac0d7ce94925f8eac367533b276d9d7db13a77ae , right before the move.

**Comment by AlekMosingiewicz** [comment]
Ok, we need to close it somehow, so I propose to:
 - keep the changes in Travis CI, with upgraded GCC compilers and XCode ( Apple ) pipeline blocked until Travis upgrades itself to include newer versions of XCode
 - remove older versions of Visual Studio, as suggested by @lefticus , from Appveyor pipeline
 - whether we keep enforcing the C++17 standard in CMake configuration is a decision I leave to @lefticus , however putting it there eliminated most of the problems with XCode build that were not directly related to missing libraries

The problems that emerged but are not directly related to pipeline itself are:
  - inability to build the project in newer versions of Visual Studio. Since it's not directly related to CI configuration I'd handle it separately
  - problems with building the project on older versions of XCode - it's related to missing libraries, but not directly to CI configuration

On my part I believe I did everything I could to fix the CI config.

**Comment by lefticus** [comment]
I'm going to go ahead and merge this and sorry for the delays!

**Comment by lefticus** [comment]
Do you mind if I do a squash merge so the history makes more sense @AlekMosingiewicz ?

**Comment by AlekMosingiewicz** [comment]
Given how little changes come out of those 31 commits it makes sense. Please go ahead!

**Comment by lefticus** [comment]
Done! I've been doing the squashes on my own changes that are all just CI as well.

**Pull request merged**